plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts

```diff
@@ -0,0 +1,236 @@
+   * @param entityRef - stringified entity ref
+   * @returns The language breakdown
+   */
+  private async getLanguagesFromLinguistAPI(
+    entityRef: string,
+  ): Promise<string[]> {
+    this.logger.debug(`Fetched languages from linguist API`, {
+      ...this.loggerMeta,
+      entityRef,
+    });
+
+    const baseUrl = this.config.getString('backend.baseUrl');
+    const linguistApi = new URL(`${baseUrl}/api/linguist/entity-languages`);
+    linguistApi.searchParams.append('entityRef', entityRef);
+    const linguistData = await fetch(linguistApi).then(
+      res => res.json() as Promise<Languages>,
+    );
+    if (!linguistData || !linguistData.processedDate) {
+      return [];
+    }
+
+    return linguistData.breakdown
+      .filter(b => b.type === 'programming' && b.bytes > this.bytesThreshold)
+      .map(b => b.name);
+  }
+
+  /**
+   * Cached wrapper around getLanguagesFromLinguistAPI
+   * @param cache - The CatalogProcessorCache
+   * @param entityRef - Stringified entity references
+   *
+   * @returns List of languages
+   */
+  private async getCachedLanguages(
+    cache: CatalogProcessorCache,
+    entityRef: string,
+  ): Promise<string[]> {
+    let cachedData = (await cache.get(entityRef)) as CachedData | undefined;
+    if (!cachedData || this.isExpired(cachedData)) {
+      const languages = await this.getLanguagesFromLinguistAPI(entityRef);
+      cachedData = { languages, cachedTime: Date.now() };
+      await cache.set(entityRef, cachedData);
+    }
+    this.logger.debug(`Fetched cached languages ${cachedData.languages}`, {
+      ...this.loggerMeta,
+      entityRef,
+    });
+    return cachedData.languages;
+  }
+
+  /**
+   * Determines if cached data is expired based on TTL
+   *
+   * @param cachedData - The cached data for this entity
+   * @returns True if data is expired
+   */
+  private isExpired(cachedData: CachedData): boolean {
+    const elapsed = Date.now() - (cachedData.cachedTime || 0);
+    return elapsed > this.cacheTTLMilliseconds;
+  }
+
+  /**
+   * This post-processor will fetch linguist data for a Component and convert the language breakdown
+   * into entity tags which will be appended to the entity.
+   *
+   * @public
+   */
+  async postProcessEntity(
+    entity: Entity,
+    _: any,
+    __: any,
+    cache: CatalogProcessorCache,
+  ): Promise<Entity> {
+    if (!this.entityValidator(entity)) {
+      return entity;
+    }
+    const entityRef = stringifyEntityRef(entity);
+    this.logger.debug(`Processing ${entityRef}`, {
+      ...this.loggerMeta,
+      entityRef,
+    });
+
+    const languages =
+      this.cacheTTLMilliseconds > 0
+        ? await this.getCachedLanguages(cache, entityRef)
+        : await this.getLanguagesFromLinguistAPI(entityRef);
+
+    const tags = (entity.metadata.tags ||= []);
+    const originalTagCount = tags.length;
+
+    languages.forEach(lang => {
+      if (!tags.includes(lang)) {
+        const cleanedUpLangTag = this.tagTransformer(lang);
```

Comment:
If I'm reading this correctly if they don't pass a `tagTranformer` function then the languages as used as is from Linguist? Won't that be problematic as the languages can have spaces in their names and I know the tags in Backstage won't allow that by default. Maybe we should have a default for this?

Here's the list of languages Linguist uses: https://github.com/github-linguist/linguist/blob/master/lib/linguist/languages.yml

Reply:
good idea, i haven't tested all of these.

Reply:
I mean, don't feel like you need to test all of them but if we sample from that list and pull out the odd ones to have in automated tests that would be good

Reply:
Just tested a few with spaces and special characters, I didn't see any problems. So not sure if there is anything to do here.

![image](https://github.com/backstage/backstage/assets/16994570/365225f0-9849-4747-9b24-9bfc76da5f4e)

Reply:
@freben, I'm happy that the above works but why does it work? The docs say: "Each tag must be sequences of [a-z0-9] separated by -, at most 63 characters in total." (https://backstage.io/docs/features/software-catalog/descriptor-format#tags-optional) Are the docs wrong and need to be updated?

Reply:
from experience, the mentioned rules get applied for tags -- at least, when provided in catalog files. I wonder why it worked here 🤔

Reply:
Could this be an "order of processors" problem? Maybe the tags don't get validated since this gets added to the catalog builder at a later stage?

Reply:
@awanlin the docs seem not the be up-to-date according to some code I found

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts#L82
```ts
          case 'isValidTag':
            expectation =
              'a string that is sequences of [a-z0-9+#] separated by [-], at most 63 characters in total';
            break;
```

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts#L148
```ts
    const tags = entity.metadata.tags ?? [];

    for (let i = 0; i < tags.length; ++i) {
      require(`tags.${i}`, tags[i], this.validators.isValidTag);
    }
```

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/validation/makeValidator.ts#L29
```ts
  isValidTag: (value: unknown): boolean => {
    // NOTE(freben): This one is a bit of an oddball and doesn't fit well anywhere to delegate to, so it's just inlined for now.
    return (
      typeof value === 'string' &&
      value.length >= 1 &&
      value.length <= 63 &&
      /^[a-z0-9:+#]+(\-[a-z0-9:+#]+)*$/.test(value)
    );
  },
```

Reply:
aha! If I switch this from post-process to pre-process, I will see validation errors.

![image](https://github.com/backstage/backstage/assets/16994570/41bb2441-f81b-4672-821a-67f3a84a4f6e)

Reply:
Yeah entity mutations should go in pre. It's a bit unfortunate that it's technically possible to break contract in post...

Reply:
I realize this puts this PR into a bit of a bind, but perhaps this is a use case for [Changing the validation rules for the entity envelope](https://backstage.io/docs/features/software-catalog/extending-the-model#changing-the-validation-rules-for-the-entity-envelope-or-metadata-fields)?

This is a plugin after all, so integrators are opting into this functionality. The setup steps could involve also relaxing the tag restriction rules, or we could just keep this as a post-process phase to ignore the restrictions.

Reply:
Ok, looking into this some more perhaps it's not as bad as I thought.

I tested the entire linguist data set using the following script

```js
const yaml = require('js-yaml');
const fs = require('fs')

async function main() {
  const langs = yaml.load(fs.readFileSync('./config.yaml'));

  const valid = [];
  const invalid = []
  const validCleaned = [];
  const invalidCleaned = [];

  Object.keys(langs).forEach(lang => {
    if ( langs[lang].type === 'programming' ) {
      if (isValidTag(lang)) {
        valid.push(lang);
      }
      else {
        invalid.push(lang);
      }
      const cleaned = cleanTag(lang);
      if (isValidTag(cleaned)) {
        validCleaned.push(cleaned);
      }
      else {
        invalidCleaned.push(cleaned);
      }
    }
  });

  const results = {
    valid: valid.length,
    invalid: invalid.length,
    validCleaned: validCleaned.length,
    invalidCleaned: invalidCleaned.length
  }

  console.log(JSON.stringify(results, null, 2));
  console.log(invalidCleaned);
}

function isValidTag (value) {
  // NOTE(freben): This one is a bit of an oddball and doesn't fit well anywhere to delegate to, so it's just inlined for now.
  return (
    typeof value === 'string' &&
    value.length >= 1 &&
    value.length <= 63 &&
    /^[a-z0-9:+#]+(\-[a-z0-9:+#]+)*$/.test(value)
  );
}

function cleanTag (value) {
  return value
    .replace(/\s+/g, '-')
    .toLowerCase();
}

main();
```

**Results**
```
{
  "valid": 9,
  "invalid": 457,
  "validCleaned": 459,
  "invalidCleaned": 7
}
[
  'asp.net',
  "cap'n-proto",
  'f*',
  'peg.js',
  "ren'py",
  'visual-basic-.net',
  'visual-basic-6.0'
]
```

Most of the issues are the fact that capital letters and spaces are not allowed in tags. I think asp.net is the only language in the short list of "still invalid" language tags that may just need a custom mapping. So that could be hard-mapped to "asp-dot-net" and the rest I think can be ignored.

Thoughts?

Reply:
you can use the invert to catch them all. And be aware that the first char is not allowed to be `-`.

```ts
value.replace(/^[^a-z0-9:+#]/, '').replace(/[^a-z0-9:+#-]+/g, '-')
```

Reply:
re your question:
In my opinion, it should be a preprocessing step that follows the restrictions for tags. Esp. if it is part of the main repo and not a separate plugin. Requiring to change the rules for the tags field will reduce the usefulness/ability to adopt this processor.

Reply:
As the docs don't match the actually allowed chars, I just created PR #18364 to fix this.